src/optimizer/plugins/SimplifyArrayIndexingOptimizer.ts

```diff
@@ -0,0 +1,213 @@
+          if (!caretPathMap.has(originalRule.path)) {
+            caretPathMap.set(originalRule.path, new Map());
+          }
+
+          const elementCaretPathMap = caretPathMap.get(originalRule.path);
+          applySoftIndexing(element, elementCaretPathMap);
+        });
+      });
+
+      removeZeroIndices(parsedPaths);
+
+      parsedPaths.forEach((parsedRule: any, ruleIndex: number) => {
+        const originalRule = def.rules[ruleIndex];
+        originalRule.path = utils.assembleFSHPath(parsedRule.path);
+        if (originalRule instanceof ExportableCaretValueRule) {
+          originalRule.caretPath = utils.assembleFSHPath(parsedRule.caretPath);
+        }
+      });
+    });
+  }
+} as OptimizerPlugin;
+
+/**
+ * Populates each PathPart in an array of PathParts with a prefix containing the assembled path up to that element
+ * @param { path: PathPart[] } parsedPath - An array of PathParts
+ */
+function addPrefixes(parsedPath: fhirtypes.PathPart[]) {
+  parsedPath.forEach((element: fhirtypes.PathPart, elementIndex: number) => {
+    // Add a prefix to the current element containing previously parsed rule elements
+    element.prefix = utils.assembleFSHPath(parsedPath.slice(0, elementIndex));
+  });
+}
+
+/**
+ * Converts numeric indeces on a PathPart into soft indexing characters
+ * @param {PathPart} element - A single element in a rules path
+ * @param {Map<string, number} pathMap - A map containing an element's name as the key and that element's updated index as the value
+ * @param {Map<string, PathPart[]} singletonArrayElements - A map containing a string unique to each element of type array as the key, and an array of PathParts as the value
+ */
+function applySoftIndexing(element: fhirtypes.PathPart, pathMap: Map<string, number>) {
+  // Must account for a pathPart's base name, prior portions of the path, as well as any slices it's contained in.
+  const mapName = `${element.prefix ?? ''}.${element.base}|${(element.slices ?? []).join('|')}`;
+  const indexRegex = /^[0-9]+$/;
+
+  const currentNumericBracket = element.brackets?.find(bracket => indexRegex.test(bracket));
+  if (!currentNumericBracket) return; // If the element is not an array, exit the function
+
+  const bracketIndex = element.brackets.indexOf(currentNumericBracket);
+
+  if (!pathMap.has(mapName)) {
+    pathMap.set(mapName, parseInt(currentNumericBracket));
+  } else {
+    const previousNumericIndex = pathMap.get(mapName);
+    if (parseInt(currentNumericBracket) === previousNumericIndex + 1) {
+      element.brackets[bracketIndex] = '+';
+      pathMap.set(mapName, previousNumericIndex + 1);
+    } else if (parseInt(currentNumericBracket) === previousNumericIndex) {
+      element.brackets[bracketIndex] = '=';
+    } else {
+      pathMap.set(mapName, parseInt(currentNumericBracket));
+    }
+  }
+}
+
+function splitOnPathPeriods(path: string): string[] {
+  return path.split(/\.(?![^\[]*\])/g); // match a period that isn't within square brackets
+}
+
```

Comment:
Could you just add a comment here indicating that this function should be removed once we update GoFSH to use a version > 1.3.1 of SUSHI?

Reply:
Sounds reasonable!